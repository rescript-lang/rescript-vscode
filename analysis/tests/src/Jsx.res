module M = {
  @react.component
  let make = (~first, ~fun="", ~second="") => React.string(first ++ fun ++ second)
}

let _ = <M first="abc" />
//       ^def

// <M second=fi
//             ^com

// <M second="abc" f
//                  ^com

// let e = <M
//           ^com

@react.component
let make = (~first) => React.string(first)

let y = 44

// <M prop={A(3)} k
//                 ^com

// <M prop=A(3) k
//               ^com

// <M prop=foo(1+2) k
//                   ^com

// <M prop=list{1,2,3} k
//                      ^com

// <M prop=<N /> k
//                ^com

// <M prop=1.5 k
//              ^com

// <M prop=0X33 k
//               ^com

// <M prop=12e+3 k
//                ^com

// <M prop='z' k
//              ^com

// <M prop=`before${foo}` k
//                         ^com

// <M prop=module(@foo Three: X_int) k
//                                    ^com

// <M prop=%bs.raw("1") k
//                       ^com

let _ = <Component />
//         ^def

module Ext = {
  @react.component @module("@material-ui/core")
  external make: (~align: string=?) => React.element = "Typography"
}

let _ = (Ext.make, Ext.makeProps)

// <Ext al
//        ^com

// <M first
//         ^com

// <M first=#a k
//              ^com

// <M first =  ?   #a k
//                     ^com

// <M>
//    ^com

module WithChildren = {
  @react.component
  let make = (~name as _: string, ~children) => <jsx> children </jsx>
}

let _ = <WithChildren name=""> <div /> </WithChildren>
// <WithChildren
//              ^com
// <WithChildren n
//                ^com

// let c : React.e
//                ^com
// let c : ReactDOMR
//                  ^com

module DefineSomeFields = {
  type r = {thisField: int, thatField: string}
  let thisValue = 10
  // let foo x = x.th
  //                 ^com
}

// let q = DefineSomeFields.
//                          ^com
// let foo x = x.DefineSomeFields.th
//                                  ^com

let _ = x => x.DefineSomeFields.thisField + DefineSomeFields.thisValue

module Outer = {
  module Inner = {
    let hello = 3
  }
}
let _ = Outer.Inner.hello

let _ =
  <div
  // x=Outer.Inner.h
  //                ^com
    name=""
  />

let _ =
  <div
  // x=Outer.Inner.
  //               ^com
    name=""
  />

let _ =
  <div
  // x=
  //   ^com
    name=""
  />

module Nested = {
  module Comp = {
    @react.component
    let make = (~name) => React.string(name)
  }
}

let _ = <Nested.Comp name="" />

// let _ = <Nested.Co name="" />
//                   ^com

// let _ = <Nested. name="" />
//                 ^com

module Comp = {
  @react.component
  let make = (~age) => React.int(age)
}

let _ = {
  <> <Comp age=34 /> </>
  //        ^hov
}

let _ = {
  <> {<> <Comp age=34 /> </>} </>
  //            ^hov
}
