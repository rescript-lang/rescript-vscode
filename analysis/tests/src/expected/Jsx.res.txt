Definition tests/src/Jsx.res 5:9
{"uri": "Jsx.res", "range": {"start": {"line": 2, "character": 6}, "end": {"line": 2, "character": 10}}}

Complete tests/src/Jsx.res 7:2
posCursor:[8:14] posNoWhite:[8:12] Found expr:[8:2->8:13]
JSX <M:[8:2->8:3] second[8:4->8:10]=...[8:11->8:13]> _children:None
posCursor:[8:14] posNoWhite:[8:12] Found expr:[8:11->8:13]
Pexp_ident fi:[8:11->8:13]
Completable: Cjsx([M], "", [second])
[{
    "label": "first",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }, {
    "label": "fun",
    "kind": 4,
    "tags": [],
    "detail": "option<string>",
    "documentation": null
  }, {
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 9:2
posCursor:[10:18] posNoWhite:[10:17] Found expr:[10:2->10:18]
JSX <M:[10:2->10:3] second[10:4->10:10]=...[10:11->10:16] f[10:17->10:18]=...[10:17->10:18]> _children:None
posCursor:[10:18] posNoWhite:[10:17] Found expr:[10:17->10:18]
Pexp_ident f:[10:17->10:18]
Completable: Cjsx([M], f, [second, f])
[{
    "label": "first",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }, {
    "label": "fun",
    "kind": 4,
    "tags": [],
    "detail": "option<string>",
    "documentation": null
  }]

Complete tests/src/Jsx.res 11:2
posCursor:[12:12] posNoWhite:[12:10] Found expr:[12:10->12:11]
JSX <M:[12:10->12:11] > _children:None
posCursor:[12:12] posNoWhite:[12:10] Found expr:[12:10->12:11]
Pexp_ident M.createElement:[12:10->12:11]
Completable: Cjsx([M], "", [])
[{
    "label": "second",
    "kind": 4,
    "tags": [],
    "detail": "option<string>",
    "documentation": null
  }, {
    "label": "first",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }, {
    "label": "fun",
    "kind": 4,
    "tags": [],
    "detail": "option<string>",
    "documentation": null
  }, {
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 18:2
posCursor:[19:17] posNoWhite:[19:16] Found expr:[19:2->19:17]
JSX <M:[19:2->19:3] prop[19:4->19:8]=...[19:10->19:14] k[19:16->19:17]=...[19:16->19:17]> _children:None
posCursor:[19:17] posNoWhite:[19:16] Found expr:[19:16->19:17]
Pexp_ident k:[19:16->19:17]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 20:2
posCursor:[21:15] posNoWhite:[21:14] Found expr:[21:2->21:15]
JSX <M:[21:2->21:3] prop[21:4->21:8]=...[21:9->21:13] k[21:14->21:15]=...[21:14->21:15]> _children:None
posCursor:[21:15] posNoWhite:[21:14] Found expr:[21:14->21:15]
Pexp_ident k:[21:14->21:15]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 22:2
posCursor:[23:19] posNoWhite:[23:18] Found expr:[23:2->23:19]
JSX <M:[23:2->23:3] prop[23:4->23:8]=...[23:9->23:17] k[23:18->23:19]=...[23:18->23:19]> _children:None
posCursor:[23:19] posNoWhite:[23:18] Found expr:[23:18->23:19]
Pexp_ident k:[23:18->23:19]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 24:2
posCursor:[25:22] posNoWhite:[25:21] Found expr:[25:2->25:22]
JSX <M:[25:2->25:3] prop[25:4->25:8]=...[25:9->25:20] k[25:21->25:22]=...[25:21->25:22]> _children:None
posCursor:[25:22] posNoWhite:[25:21] Found expr:[25:21->25:22]
Pexp_ident k:[25:21->25:22]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 26:2
posCursor:[27:16] posNoWhite:[27:15] Found expr:[27:2->27:16]
JSX <M:[27:2->27:3] prop[27:4->27:8]=...[27:10->27:14] k[27:15->27:16]=...[27:15->27:16]> _children:None
posCursor:[27:16] posNoWhite:[27:15] Found expr:[27:15->27:16]
Pexp_ident k:[27:15->27:16]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 28:2
posCursor:[29:14] posNoWhite:[29:13] Found expr:[29:2->29:14]
JSX <M:[29:2->29:3] prop[29:4->29:8]=...[29:9->29:12] k[29:13->29:14]=...[29:13->29:14]> _children:None
posCursor:[29:14] posNoWhite:[29:13] Found expr:[29:13->29:14]
Pexp_ident k:[29:13->29:14]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 30:2
posCursor:[31:15] posNoWhite:[31:14] Found expr:[31:2->31:15]
JSX <M:[31:2->31:3] prop[31:4->31:8]=...[31:9->31:13] k[31:14->31:15]=...[31:14->31:15]> _children:None
posCursor:[31:15] posNoWhite:[31:14] Found expr:[31:14->31:15]
Pexp_ident k:[31:14->31:15]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 32:2
posCursor:[33:16] posNoWhite:[33:15] Found expr:[33:2->33:16]
JSX <M:[33:2->33:3] prop[33:4->33:8]=...[33:9->33:14] k[33:15->33:16]=...[33:15->33:16]> _children:None
posCursor:[33:16] posNoWhite:[33:15] Found expr:[33:15->33:16]
Pexp_ident k:[33:15->33:16]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 34:2
posCursor:[35:14] posNoWhite:[35:13] Found expr:[35:2->35:14]
JSX <M:[35:2->35:3] prop[35:4->35:8]=...[35:9->35:12] k[35:13->35:14]=...[35:13->35:14]> _children:None
posCursor:[35:14] posNoWhite:[35:13] Found expr:[35:13->35:14]
Pexp_ident k:[35:13->35:14]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 36:2
posCursor:[37:25] posNoWhite:[37:24] Found expr:[37:2->37:25]
JSX <M:[37:2->37:3] prop[37:4->37:8]=...[37:9->37:23] k[37:24->37:25]=...[37:24->37:25]> _children:None
posCursor:[37:25] posNoWhite:[37:24] Found expr:[37:24->37:25]
Pexp_ident k:[37:24->37:25]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 38:2
posCursor:[39:36] posNoWhite:[39:35] Found expr:[39:2->39:36]
JSX <M:[39:2->39:3] prop[39:4->39:8]=...[39:9->39:34] k[39:35->39:36]=...[39:35->39:36]> _children:None
posCursor:[39:36] posNoWhite:[39:35] Found expr:[39:35->39:36]
Pexp_ident k:[39:35->39:36]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 40:2
posCursor:[41:23] posNoWhite:[41:22] Found expr:[41:2->41:23]
JSX <M:[41:2->41:3] prop[41:4->41:8]=...[41:9->41:21] k[41:22->41:23]=...[41:22->41:23]> _children:None
posCursor:[41:23] posNoWhite:[41:22] Found expr:[41:22->41:23]
Pexp_ident k:[41:22->41:23]
Completable: Cjsx([M], k, [prop, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Definition tests/src/Jsx.res 43:11
{"uri": "Component.res", "range": {"start": {"line": 1, "character": 4}, "end": {"line": 1, "character": 8}}}

Complete tests/src/Jsx.res 52:2
posCursor:[53:8] posNoWhite:[53:7] Found expr:[53:2->53:8]
JSX <Ext:[53:2->53:5] al[53:6->53:8]=...[53:6->53:8]> _children:None
posCursor:[53:8] posNoWhite:[53:7] Found expr:[53:6->53:8]
Pexp_ident al:[53:6->53:8]
Completable: Cjsx([Ext], al, [al])
[{
    "label": "align",
    "kind": 4,
    "tags": [],
    "detail": "option<string>",
    "documentation": null
  }]

Complete tests/src/Jsx.res 54:2
posCursor:[55:10] posNoWhite:[55:8] Found expr:[55:2->55:9]
JSX <M:[55:2->55:3] first[55:4->55:9]=...[55:4->55:9]> _children:None
posCursor:[55:10] posNoWhite:[55:8] Found expr:[55:4->55:9]
Pexp_ident first:[55:4->55:9]
Completable: Cjsx([M], "", [first])
[{
    "label": "second",
    "kind": 4,
    "tags": [],
    "detail": "option<string>",
    "documentation": null
  }, {
    "label": "fun",
    "kind": 4,
    "tags": [],
    "detail": "option<string>",
    "documentation": null
  }, {
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 56:2
posCursor:[57:14] posNoWhite:[57:13] Found expr:[57:2->57:14]
JSX <M:[57:2->57:3] first[57:4->57:9]=...[57:10->57:12] k[57:13->57:14]=...[57:13->57:14]> _children:None
posCursor:[57:14] posNoWhite:[57:13] Found expr:[57:13->57:14]
Pexp_ident k:[57:13->57:14]
Completable: Cjsx([M], k, [first, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 58:2
posCursor:[59:21] posNoWhite:[59:20] Found expr:[59:2->59:21]
JSX <M:[59:2->59:3] first[59:4->59:9]=...[59:17->59:19] k[59:20->59:21]=...[59:20->59:21]> _children:None
posCursor:[59:21] posNoWhite:[59:20] Found expr:[59:20->59:21]
Pexp_ident k:[59:20->59:21]
Completable: Cjsx([M], k, [first, k])
[{
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

Complete tests/src/Jsx.res 60:2
posCursor:[61:4] posNoWhite:[61:3] Found expr:[61:2->65:69]
Pexp_apply ...[63:20->63:21] (...[61:2->63:19], ...[64:2->65:69])
posCursor:[61:4] posNoWhite:[61:3] Found expr:[61:2->63:19]
JSX <M:[61:2->61:3] > _children:61:3
posCursor:[61:4] posNoWhite:[61:3] Found expr:[61:3->63:20]
Pexp_construct :::[63:0->63:20] [63:0->63:20]
posCursor:[61:4] posNoWhite:[61:3] Found expr:__ghost__[61:3->63:20]
Pexp_construct []:__ghost__[61:3->63:20] None
[]

Complete tests/src/Jsx.res 68:2
posCursor:[69:15] posNoWhite:[69:13] Found expr:[69:2->69:14]
JSX <WithChildren:[69:2->69:14] > _children:None
posCursor:[69:15] posNoWhite:[69:13] Found expr:[69:2->69:14]
Pexp_ident WithChildren.createElement:[69:2->69:14]
Completable: Cjsx([WithChildren], "", [])
[{
    "label": "name",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }, {
    "label": "children",
    "kind": 4,
    "tags": [],
    "detail": "React.element",
    "documentation": null
  }, {
    "label": "key",
    "kind": 4,
    "tags": [],
    "detail": "string",
    "documentation": null
  }]

